### ISSUE TITLE
[syncfusion_flutter_datagrid] UI Filtering and Widget cells

### ISSUE THREAD

**Issue by divan** [syncfusion/flutter-widgets#1042]
Hi,

first of all, thank you for the incredible plugin.

I've encountered the issue when changed cells values from simple strings/ints to the widgets. I.e. instead of nation name, I want to show flag and the name. When I use Widgets as cells values, UI filtering seems to break:

![Screenshot 2023-01-20 at 00 41 55](https://user-images.githubusercontent.com/880202/213578482-54024814-8fb9-4c36-be06-6a203e19e718.png)

Basically when I construct rows, I set:

```dart
                return DataGridCell<dynamic>(
                  columnName: key,
                  value: NationCell(value),
                );
```

NationCell is just a simple Row widget.

Is there any way to separate the cell value from cell representation? Or tell UI filtering code what's the underlying value is?

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @divan,

You can load the country name with the flag as the widget in the buildRow method for the corresponding column. In the underlying collection i.e., DataGridCell, you can have only the country name, this way the filter UI will display the country name. Filter UI will only display the items from the DataGridCell.value property which is maintained in the DataGridSource.rows property. 

I have provided a sample that illustrates this solution, please refer to the attached sample and the following code snippet for more information.

```dart
class EmployeeDataSource extends DataGridSource {
  
  …

  void buildDataGridRows(List<Employee> employees) {
    datagridRow = employees
        .map<DataGridRow>((employee) => DataGridRow(cells: [
              DataGridCell<int>(columnName: 'id', value: employee.id),
              DataGridCell<String>(columnName: 'name', value: employee.name),
              DataGridCell<String>(
                  columnName: 'country', value: employee.country),
              DataGridCell<String>(
                  columnName: 'designation', value: employee.designation),
              DataGridCell<int>(columnName: 'salary', value: employee.salary),
            ]))
        .toList();
  }

  @override
  DataGridRowAdapter buildRow(DataGridRow row) {
    return DataGridRowAdapter(
        cells: row.getCells().map<Widget>((dataGridCell) {
      return dataGridCell.columnName == 'country'
          ? Row(
              children: [
                Image.asset(
                  'assets/flag/${dataGridCell.value}.png',
                  height: 20,
                ),
                const Padding(padding: EdgeInsets.all(10)),
                Text(dataGridCell.value.toString())
              ],
            )
          : Container(
              alignment: Alignment.center,
              padding: const EdgeInsets.symmetric(horizontal: 8.0),
              child: Text(dataGridCell.value.toString()),
            );
    }).toList());
  }
}

List<Employee> populateData() {
  return [
    Employee(10001, 'Jack', 'UK', 'Manager', 100000),
    Employee(10002, 'Kathryn', 'Brazil', 'Project Lead', 70000),
    Employee(10003, 'Lara', 'Mexico', 'Developer', 55000),
    Employee(10004, 'Michael', 'France', 'Developer', 53000),
    Employee(10005, 'Williams', 'Germany', 'Developer', 50000),
    Employee(10006, 'Ellis', 'Argentina', 'Designer', 40000),
    Employee(10007, 'James', 'Canada', 'Developer', 44000),
    Employee(10008, 'Perry', 'Spain', 'Developer', 42000),
    Employee(10009, 'Gable', 'Italy', 'Support', 38000),
    Employee(10010, 'Crowley', 'Portugal', 'Administrator', 33000)
  ];
}

class Employee {
  Employee(this.id, this.name, this.country, this.designation, this.salary);

  final int id;
  final String name;
  final String country;
  final String designation;
  final int salary;
}

```

**Sample Link: https://www.syncfusion.com/downloads/support/directtrac/general/ze/Sample1209120807**



![image](https://user-images.githubusercontent.com/93904422/213668305-f1037874-d95b-49e8-a95d-8890deb39454.png)


Regards,
Tamilarasan

**Comment by divan** [comment]
@Tamilarasan-Paranthaman thank you for the prompt and elaborate reply!

Interesting! That makes sense. In my case I have a wrapper around DataGrid that tries to be smart about data types of the cells, but I see that this approach won't work properly.

Anyway, you answer is super helpful, thanks again!

**Comment by divan** [comment]
@Tamilarasan-Paranthaman I'll reopen this as I have a follow up question, but want to be clear that my original problem was solved with your reply.

The follow up question is this - is this possible to use original data in `buildRow` instead of cell value? For example, I have a list of objects of type `Nation`:

```dart
class Nation {
    final IconData flag;
    final String name;
...
}
```

and I use object of type `Nation` instead of just string with the country name. In the widget for the cell that represents country I want to show flag icon based on `Nation.flag` value, and the text based on `Nation.name` value. How can I do this?

In the proposed solution the assumption is that the value is first converted to the comparable string (or some other basic type like int), and then based on this converted value (stored in DataGridCell.value) we build our widget. But I see many cases where it won't be enough, and access to the original object is needed to build a cell widget.

Any suggestions for this?

**Comment by Tamilarasan-Paranthaman** [comment]
@divan,

To meet your requirement, you can override the **toString()** method in the **Nation** class. This way, the name value can be returned when the toString() method is called in the filter popup, enabling the desired behavior. We have created a sample that demonstrates this approach, and you can find the relevant code snippet below.

```dart
class EmployeeDataSource extends DataGridSource {

…

void buildDataGridRows(List<Employee> employees) {
    datagridRow = employees
        .map<DataGridRow>((employee) => DataGridRow(cells: [
              DataGridCell<int>(columnName: 'id', value: employee.id),
              DataGridCell<String>(columnName: 'name', value: employee.name),
              DataGridCell<Nation>(
                  columnName: 'country', value: employee.nation),
              DataGridCell<String>(
                  columnName: 'designation', value: employee.designation),
              DataGridCell<int>(columnName: 'salary', value: employee.salary),
            ]))
        .toList();
  }

  @override
  DataGridRowAdapter buildRow(DataGridRow row) {
    return DataGridRowAdapter(
        cells: row.getCells().map<Widget>((dataGridCell) {
      return dataGridCell.columnName == 'country'
          ? Row(
              children: [
                dataGridCell.value.flag,
                const Padding(padding: EdgeInsets.all(10)),
                Text(dataGridCell.value.name)
              ],
            )
          : Container(
              alignment: Alignment.center,
              padding: const EdgeInsets.symmetric(horizontal: 8.0),
              child: Text(dataGridCell.value.toString()),
            );
    }).toList());
  }
}

class Employee {
  Employee(this.id, this.name, this.nation, this.designation, this.salary);

  final int id;
  final String name;
  final Nation nation;
  final String designation;
  final int salary;
}

```
**In Nation class:**
```dart
class Nation {
  Nation(this.flag, this.name);
  final Image flag;
  final String name;

@override
  String toString() {
    return name;
  }
}

```
**Sample: https://www.syncfusion.com/downloads/support/directtrac/general/ze/sample1932785739**


![image](https://user-images.githubusercontent.com/93904422/219286965-9acc401d-3ffa-49ce-bdd5-612f3de1f5c1.png)

Regards,
Tamilarasan

**Comment by divan** [comment]
@Tamilarasan-Paranthaman oh wow, how did I miss that. Right, it makes sense, though it didn't work for me properly until I implemented equality operator (==) and hashCode for that class.

So to dig this topic of value/representation separation - how would it work if, say, it's an integer value? Quick example use case - we have class `Age()`, which has one propery `int age`. It would be nice to have the numeric sort and advanced filters provided by this package (i.e. "filter values less than"), but I want custom representation through custom widget AgeCell() (let's say it just displays string "Child", "Teen", "Adult" in different colors).

`.toString()` would just convert value into the string, which kinda works, but looses the benefits of telling DataGrid that it's an int value. I think string value will work for many cases, but still interesting how shall it work with non-string values.

And thanks again for the fantastic responses. I rarely see so much quality in responses and support as with SyncFusion.

**Comment by Tamilarasan-Paranthaman** [comment]
@divan,

Based on the dart language support, it is not possible to achieve your requirement using non-string data types. i.e., considering the custom type object as integer. If you attempt to use a non-string value with this workaround, the DataGrid will consider it as a string, and the filtering options will be displayed accordingly.

**Comment by Trung15010802** [comment]
But what if in another column we still need a Object with different property? How to overides toString() for multiple column?
